### PR TITLE
[oauth] Add support for custom deserialization of AccessTokenResponse

### LINKED ISSUE #1888: [oauth] scope in AccessTokenResponse
Hello,
I'm reviewing the netatmo binding and would like to remove all external dependencies. Thus, using OAuthFactory.
This works nearly well, except that Netatmo API response is formed this way : 
````
{
"access_token":"5xxxxx95e984",
"refresh_token":"51cyyyyy841a489a4a346eb3b",
"scope":["read_station"],
"expires_in":10800,
"expire_in":10800
}
````
While in the API documentation, the scope should not be part of the answer, even worth, its a string array instead of the expected string.
So OAuthConnector fails to deserialize the AccessTokenResponse.

While on one hand, I can ask Netatmo (that I did) to respect better the RFC, I'm quite realistic regarding my chance to succeed.
Does anybody have an idea on how I could make this foot fit in the shoe ?

### PR DESCRIPTION
**Pull request by jlaur**
This reverts #1891 and fixes #1888 by providing a new way to customize deserialization of `AccessTokenResponse` which has been tested with the Netatmo binding. See openhab/openhab-addons#14755 for a description of the concrete Netatmo problem that so far prevented Netatmo from using the core OAuth implementation.

In this new approach a custom `GsonBuilder` can be provided which can be preconfigured with custom type adapters etc.

This fixes the issue with #1891:
```
[oauth2client.internal.OAuthConnector] - Unable to construct custom deserializer 'org.openhab.binding.netatmo.internal.deserialization.AccessTokenResponseDeserializer'
java.lang.ClassNotFoundException: org.openhab.binding.netatmo.internal.deserialization.AccessTokenResponseDeserializer cannot be found by org.openhab.core.auth.oauth2client_4.0.0.202304080304
```

but also adds more flexibility, for example by allowing multiple type adapters, not only a single fixed deserialization class.

### CODE REVIEW

File: bundles/org.openhab.core.auth.oauth2client/src/main/java/org/openhab/core/auth/oauth2client/internal/OAuthClientServiceImpl.java
Comment (jlaur) [inline bundles/org.openhab.core.auth.oauth2client/src/main/java/org/openhab/core/auth/oauth2client/internal/OAuthClientServiceImpl.java:416]: This is not the factory pattern, just a method named similarly, but actually creating a new object once again. This could probably be moved to `OAuthFactory`.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/auth/client/oauth2/OAuthClientService.java
Comment (jlaur) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/auth/client/oauth2/OAuthClientService.java:299]: There are no usages of this method in the addons repository.

**Review by clinique — APPROVED** [review]
LGTM

**Comment by lolodomo** [comment]
@J-N-K @kaikreuzer : there might be not real risk to merge that PR inside milestone 2.
@jlaur : do you think there is any risk to break something?

**Comment by jlaur** [comment]
> do you think there is any risk to break something?

Not really, I'm removing something which is not used and which doesn't work. The new approach was on purpose introduced in the same fashion as the previous one which was already accepted and thought through. The client service objects are cached without the custom builder, which is always added afterwards in the newly created object, so it should not break the main path. It would say it doesn't change anything radically for better or worse.

I have tested the new method with openhab/openhab-addons#14780 and it works as intended. It would be cool to have it, so I can finish that PR the day after when a new snapshot is available.

Anyway, I will leave it to the @openhab/core-maintainers to assess the risk. 🙂

**Comment by J-N-K** [comment]
Why do we use `GsonBuilder` instead of adding multiple `JsonDeserializer`s? E.g. in the `ThingBuilder` we can call `.withChannel` several times and all of them get added.

**Comment by jlaur** [comment]
> Why do we use `GsonBuilder` instead of adding multiple `JsonDeserializer`s? E.g. in the `ThingBuilder` we can call `.withChannel` several times and all of them get added.

Please see the description. The approach of providing a deserializer directly didn't work. I went with the `GsonBuilder` approach to work around that issue, but also to provide a more flexible approach giving the caller much more control since the builder can be freely configured with both serializers, deserializers etc.

**Review by J-N-K — APPROVED** [review]

**Pull request merged**
